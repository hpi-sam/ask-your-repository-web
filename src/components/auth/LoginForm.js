// @flow
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import Button from 'react-validation/build/button';
import { login } from '../../state/auth/auth.actionCreators';
import './LoginForm.scss';

type Props = {
  dispatch: Function,
}

type State = {
  email: string,
  password: string,
}

class LoginForm extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      email: '',
      password: '',
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleSubmit = (e) => {
    e.preventDefault();

    const { email, password } = this.state;
    const { dispatch } = this.props;
    if (email && password) {
      dispatch(login(email, password));
    }
  }

  render() {
    const { email, password } = this.state;
    return (
      <Form onSubmit={this.handleSubmit} className="LoginForm">
        <h1>Login</h1>
        <div className="form-input">
          <label>
            Email or Username:
            <Input
              type="text"
              name="email"
              value={email}
              onChange={this.handleChange}
              data-cy="login-email-input"
            />
          </label>
        </div>
        <div className="form-input">
          <label>
            Password:
            <Input
              type="password"
              name="password"
              value={password}
              onChange={this.handleChange}
              data-cy="login-password-input"
            />
          </label>
        </div>
        <div>
          <Link to="/register">No account yet? Register here.</Link>
        </div>
        <div>
          <Button data-cy="login-submit-button">
            Submit
          </Button>
        </div>
      </Form>
    );
  }
}

export default connect()(LoginForm);
