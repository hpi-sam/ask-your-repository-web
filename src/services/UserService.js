// @flow
import humps from 'humps';
import api from '../config/api';
import type { User, UserParams } from '../models/User';

class UserService {
  static async get(id: string): Promise<User> {
    const response = await api.get(`/users/${id}`);
    return response.data;
  }

  static async list(): Promise<User[]> {
    const response = await api.get('/users', {});

    return response.data;
  }

  static async create(formData: UserParams): Promise<User> {
    try {
      const response = await api.post('/users', formData);
      return response.data;
    } catch (e) {
      throw e;
    }
  }

  static async login(emailOrUsername: string, password: string): Promise<User> {
    try {
      const response = await api.post('/authentications', humps.decamelizeKeys({ emailOrUsername, password }));
      return response.data;
    } catch (e) {
      throw e;
    }
  }

  static async logout() {
    try {
      await api.delete('/authentications');
    } finally {
      localStorage.removeItem('user');
    }
  }
}

export default UserService;
