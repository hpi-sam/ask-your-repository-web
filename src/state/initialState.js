// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
