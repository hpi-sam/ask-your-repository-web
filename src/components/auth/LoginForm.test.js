// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import LoginForm from './LoginForm';
import UserService from '../../services/UserService';
// import UserFactory from '../../factories/UserFactory';
import initialState from '../../state/initialState';

const mockStore = configureMockStore();

jest.mock('../../services/UserService');

describe('<LoginForm />', () => {
  let wrapper;
  let store;

  beforeAll(() => {
    UserService.logout.mockImplementation(() => Promise.resolve());
  });

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount((
      <Provider store={store}>
        <MemoryRouter>
          <LoginForm />
        </MemoryRouter>
      </Provider>
    )).find('LoginForm');
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
