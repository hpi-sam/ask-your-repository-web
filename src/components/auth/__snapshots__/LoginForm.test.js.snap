// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> renders correctly 1`] = `
<LoginForm
  dispatch={[Function]}
>
  <Form(t)
    className="LoginForm"
    onSubmit={[Function]}
  >
    <t
      className="LoginForm"
      getValues={[Function]}
      hideError={[Function]}
      onSubmit={[Function]}
      showError={[Function]}
      validate={[Function]}
      validateAll={[Function]}
    >
      <form
        className="LoginForm"
        onSubmit={[Function]}
      >
        <h1>
          Login
        </h1>
        <div
          className="form-input"
        >
          <label>
            Email or Username:
            <Control(s)
              data-cy="login-email-input"
              name="email"
              onChange={[Function]}
              type="text"
              validations={Array []}
              value=""
            >
              <s
                data-cy="login-email-input"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value=""
              >
                <div>
                  <input
                    data-cy="login-email-input"
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
              </s>
            </Control(s)>
          </label>
        </div>
        <div
          className="form-input"
        >
          <label>
            Password:
            <Control(s)
              data-cy="login-password-input"
              name="password"
              onChange={[Function]}
              type="password"
              validations={Array []}
              value=""
            >
              <s
                data-cy="login-password-input"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value=""
              >
                <div>
                  <input
                    data-cy="login-password-input"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="password"
                    value=""
                  />
                </div>
              </s>
            </Control(s)>
          </label>
        </div>
        <div>
          <Link
            replace={false}
            to="/register"
          >
            <a
              href="/register"
              onClick={[Function]}
            >
              No account yet? Register here.
            </a>
          </Link>
        </div>
        <div>
          <Button(s)
            data-cy="login-submit-button"
          >
            <s
              data-cy="login-submit-button"
              hasErrors={false}
            >
              <button
                data-cy="login-submit-button"
                disabled={false}
              >
                Submit
              </button>
            </s>
          </Button(s)>
        </div>
      </form>
    </t>
  </Form(t)>
</LoginForm>
`;
